rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -133,96 +112,43 @@ public static boolean isNotPartOfRunningBuild(FlowExecution execution) {
-                    endTime = TimingAction.getStartTime(flowEndNode);
+        FlowExecution exec = node.getExecution();
+        FlowNode nextNode = null;
 
-                    // If the node is running then we might want to use the "now" time as the end time.
-                    // Otherwise we are using the start time of the node that is running, which is not
-                    // changing i.e. will look as though the node is not running.
-                    if (flowEndNode.isRunning() && !execution.isComplete()) {
-                        // But only do this if the node is not paused e.g. for input.
-                        if (!isPauseNode(flowEndNode)) {
-                            long currentTime = System.currentTimeMillis();
-                            if (currentTime > endTime) {
-                                endTime = currentTime;
-                            }
-                        }
-                    }
-                }
-
-                ExecDuration execDuration = new ExecDuration();
-                if (startTime != 0) {
-                    execDuration.setTotalDurationMillis(endTime - startTime);
-
-                    // Calculate the stage pause duration.
-                    int nextStageIndex;
-                    if (nextStageNode != null) {
-                        nextStageIndex = findStageStartNodeIndex(allNodesSorted, nextStageNode);
-                    } else {
-                        nextStageIndex = allNodesSorted.size(); // +1 from end of list
-                    }
-                    for (int i = stageStartNodeIndex; i < nextStageIndex; i++) {
-                        FlowNode node = allNodesSorted.get(i);
-                        execDuration.setPauseDurationMillis(execDuration.getPauseDurationMillis() + PauseAction.getPauseDuration(node));
-                    }
+        // Look for the next node or the one after it to see if it follows the current node, this can be much faster
+        try {
+             nextNode = exec.getNode(Integer.toString(iota + 1));
+            if (nextNode != null && nextNode.getParents().contains(node)) {
+                return nextNode;
+            }
+        } catch (IOException ioe) {
+            try {
+                nextNode = exec.getNode(Integer.toString(iota + 2));
+                if (nextNode != null && nextNode.getParents().contains(node)) {
+                    return nextNode;
                 }
-                return execDuration;
+            } catch (IOException ioe2) {
+                // Nope miss again, let's do the harder thing.
             }
         }
 
-        return new ExecDuration();
+        // Find node after this one, scanning everything until this one
+        final FlowNode after = new ForkScanner().findFirstMatch(node.getExecution().getCurrentHeads(), Collections.singletonList(node), new Predicate<FlowNode>() {
+            public boolean apply(@Nonnull FlowNode f) {
```

Comment:
No, the interface method is `@Nullable`, so it is up to the method accepting the `Predicate` to document its nullness semantics. In this case I think the expectation is that the argument should never be null.